### PR TITLE
Cache rmeta files whenever metadata is present in emit.

### LINKED ISSUE #429: Fail to build after new rustc pipelining compilation feature
### Basic information

sccache version: 0.2.8
OS: Debian GNU/Linux 9.9 (stretch) (4.9.0-8-amd64 #1 SMP Debian 4.9.144-3.1 (2019-02-19) x86_64 GNU/Linux)
cargo version: cargo 1.36.0-nightly (c4fcfb725 2019-05-15)
rustc version: rustc 1.36.0-nightly (6afcb5628 2019-05-19)

### Reproducible steps

```bash
% rm -rf ~/.cache/sccache
% export RUSTC_WRAPPER=sccache
% export CARGO_BUILD_PIPELINING=true 
% git clone https://github.com/m4b/goblin.git
% cd goblin
% cargo build
% cargo clean
% cargo build
   Compiling proc-macro2 v0.4.30
   Compiling semver-parser v0.7.0
   Compiling unicode-xid v0.1.0
   Compiling syn v0.15.34
   Compiling cfg-if v0.1.9
   Compiling plain v0.2.3
   Compiling log v0.4.6
sccache: encountered fatal error
sccache: error : Failed to open file for hashing: "/home/lzutao/forked/rust/goblin/target/debug/deps/libcfg_if-e6b10e8042cb123a.rmeta"
sccache:  cause: Failed to open file for hashing: "/home/lzutao/forked/rust/goblin/target/debug/deps/libcfg_if-e6b10e8042cb123a.rmeta"
sccache:  cause: No such file or directory (os error 2)
error: Could not compile `log`.

Caused by:
  process didn't exit successfully: `sccache rustc --crate-name log /home/lzutao/.cargo/registry/src/github.com-1ecc6299db9ec823/log-0.4.6/src/lib.rs --color always --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 -C metadata=aa0fc3947c9b41c0 -C extra-filename=-aa0fc3947c9b41c0 --out-dir /home/lzutao/forked/rust/goblin/target/debug/deps -L dependency=/home/lzutao/forked/rust/goblin/target/debug/deps --extern cfg_if=/home/lzutao/forked/rust/goblin/target/debug/deps/libcfg_if-e6b10e8042cb123a.rmeta --cap-lints allow --error-format=json --json-rendered=termcolor -Zunstable-options -Zemit-artifact-notifications` (exit code: 254)
warning: build failed, waiting for other jobs to finish...
error: build failed
```

### LINKED ISSUE #436: unconditionally rebuilds all deps on rust nightly
calling `cargo build ; cargo build` will cause sccache to try to rebuild the entire dep tree + root crate twice, even though nothing changed and it should technically not.

````
cargo clean ; cargo build ; cargo build
   Compiling serde v1.0.80
   Compiling ryu v0.2.7
   Compiling itoa v0.4.3
   Compiling serde_json v1.0.39 (/home/matthias/vcs/github/serde_json)
    Finished dev [unoptimized + debuginfo] target(s) in 3.75s
   Compiling itoa v0.4.3
   Compiling ryu v0.2.7
   Compiling serde v1.0.80
   Compiling serde_json v1.0.39 (/home/matthias/vcs/github/serde_json)
    Finished dev [unoptimized + debuginfo] target(s) in 1.16s
````

Cargo with rustc (and no RUSTC_WRAPPER) gets this right and does not rebuild anything on the second "cargo build" run.

This happens since this commit in the cargo repo: https://github.com/rust-lang/cargo/commit/6b28a0c0505f19a574a44bffdeccbc07a2d25980

Related cargo ticket: https://github.com/rust-lang/cargo/issues/6976

### PR DESCRIPTION
**Pull request by chmanchester**
Fixes #436, Fixes #429

### CODE REVIEW

File: src/compiler/rust.rs [hunk 1]
```diff
@@ -915,7 +916,7 @@ fn parse_arguments(arguments: &[OsString], cwd: &Path) -> CompilerArguments<Pars
         staticlibs: staticlibs,
         crate_name: crate_name.to_string(),
         dep_info: dep_info.map(|s| s.into()),
-        rename_rlib_to_rmeta,
+        emit: emit,
```
Comment (luser) [inline src/compiler/rust.rs]:
```suggestion
        emit,
```

File: src/compiler/rust.rs [hunk 2]
```diff
@@ -1064,7 +1065,26 @@ impl<T> CompilerHasher<T> for RustHasher
...
+                    let rlibs: HashSet<_> = outputs.iter().cloned().filter(|p| {
+                        p.ends_with(".rlib")
+                    }).collect();
+                    for lib in rlibs {
+                        let rmeta = lib.replacen(".rlib", ".rmeta", 1);
+                        // Do this defensively for future versions of rustc that may
+                        // be fixed.
+                        if !outputs.contains(&rmeta) {
+                            outputs.push(rmeta);
+                        };
```
Comment (luser) [inline src/compiler/rust.rs]:
```suggestion
                        }
```

File: src/compiler/rust.rs [hunk 3]
```diff
@@ -1064,7 +1065,26 @@ impl<T> CompilerHasher<T> for RustHasher
...
+                        if !emit.contains("link") {
+                            outputs.retain(|p| *p != lib);
+                        };
```
Comment (luser) [inline src/compiler/rust.rs]:
```suggestion
                        }
```

**Comment by theduke** [comment]
Thanks, this was driving me crazy!

**Pull request merged**